lib/wikitrivia/game.ex

```diff
@@ -1,100 +1,229 @@
-    Agent.update(agent_name_by_game_id(game_id), add_player(player_name))
+  def add_player(game_id, player_id) do
+    GenServer.call(name_by_game_id(game_id), {:add_player, player_id})
   end
 
-  def award_points(game_id, player_name, points) do
-    Agent.update(agent_name_by_game_id(game_id), award_points(player_name, points))
+  def update_player_name(game_id, player_id, player_name) do
+    GenServer.call(name_by_game_id(game_id), {:update_player_name, player_id, player_name})
   end
 
-  # Starts the pose question -> show results loop. Ends after num_questions_left reaches 0.
-  # game_timer_callback function will be called after every timed state change.
+  def answer_question(game_id, player_id, answer) do
+    GenServer.call(name_by_game_id(game_id), {:answer_question, player_id, answer, Time.utc_now()})
+  end
+
+  # TODO: We'll probably want to alter the behavior here s.t. the final player answering the
+  # question immediately triggers the next phase (the question_results) phase. (there's no reason
+  # to wait those extra seconds, since you can't change your answer - OR maybe we let people change
+  # their answers after everyone has answered).
+
+  # Starts the pose question -> show results loop. Ends after current_question reaches
+  # num_questions - 1. The provided game_timer_callback function will be called after every timed
+  # state change.
+
+  # THIS EVENT CALL DOES NOT RETURN ANY STATE (the provided callback function should be relied upon for that)
   def start(game_id, game_timer_callback) do
-    start_question(game_id, game_timer_callback)
+    GenServer.cast(name_by_game_id(game_id), {:start_game, game_timer_callback})
   end
 
-  # State structure of a Game. Details:
-  #   * players - Set of all players in this game.
-  #   * scores - Map of all player scores in this game, keyed by player name
-  #   * num_questions_left - Number of questions remaining to answer in this game
-  #   * timer_state - Current state of timed component of this game. :off before entering timed
-  #       periods, otherwise the name of the period (:question, :question_results, :done)
-  #   * timer_data - Data associated with the timer_state (e.g. the question)
-  defp default_state do
-    %{
-      players: MapSet.new(),
-      scores: %{},
-      num_questions_left: 5,
-      timer_state: :off,
-      timer_data: %{}
-    }
+
+  ### Server (callbacks) ###
+
+  def init(state) do
+    {:ok, state}
   end
 
-  defp add_player(player_name) do
-    fn (state = %{players: players, scores: scores}) ->
-      players = players |> MapSet.put(player_name)
-      scores = scores |> Map.put(player_name, 0)
-      %{state | players: players, scores: scores}
+  def handle_call(:get_state, _from, state), do: {:reply, state, state}
+
+  def handle_call({:add_player, player_id}, _from, state = %{players: players, scores: scores}) do
+    if Map.has_key?(players, player_id) do
+      {:reply, {:no_change, state}, state}
```

Comment:
What is the use case for responding with `:no_change`?

Reply:
@denheck I was figuring this information could be passed along to the client to avoid unnecessary re-renders, and the state management system would deal with deciding whether or not to take an action given some input. So instead of this (current code in channel):

```elixir
    # Remove state check from channel code
    player_id = if Game.has_player(game_id, player_id) do
      player_id
    else
      Game.add_player(game_id, player_id)
    end
```

We'd do something like this:
```elixir
result = Game.add_player(game_id, player_id) # knows whether or not to no-op
case result do
  {:ok, ...} -> whatever
  {:no_change, ...} -> whatever
```

If we don't care about having rerenders even if the state doesn't change, and it's adding too much complexity, we can take it out and always return the "new state" instead (even if it's the same as the old state)